### PR TITLE
Fix index_reduce sampler filter when op_info.variant_test_name is specified

### PR DESCRIPTION
**Pull request by pearu**
As in the title: `index_reduce` sample must correspond to reduction type specified by `variant_test_name`.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122579
* __->__ #123375

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire

### CODE REVIEW

**Review by zou3519 — APPROVED** [review]

File: test/inductor/test_torchinductor_opinfo.py [hunk 1]
```diff
@@ -517,6 +507,9 @@ def fn(*args, **kwargs):
             # but that when we do backwards we expect other ops like add to work
             and not dtype == torch.complex32
         )
+        if op_name != "index_reduce.prod":
+            requires_grad = False
```
Comment (peterbell10) [inline test/inductor/test_torchinductor_opinfo.py:511]: Why this change when it seemed to work before?

**Review by peterbell10 — CHANGES_REQUESTED** [review]

File: test/inductor/test_torchinductor_opinfo.py (hunk 1 above)
Comment (pearu) [reply test/inductor/test_torchinductor_opinfo.py:511]: It didn't work before.

Before this PR, reduction types other that `prod` were never nested in the `test_comprehensive` test because this test uses only the first sample from an operation sample factory unless the operation is listed in the `inductor_all_samples` list, and `index_reduce` was not listed there.

When removing the if-block, the following tests fail:
```
test_comprehensive_index_reduce_mean_cpu_float64
test_comprehensive_index_reduce_amax_cuda_float16
test_comprehensive_index_reduce_amax_cuda_float32
test_comprehensive_index_reduce_amin_cuda_float64
test_comprehensive_index_reduce_cuda_float64
```
because tensors/scalars are not close. While the first test can be fixed by adjusting atol/rtol values, the other tests fail because the actual gradient values are incorrect. Disabling these tests at opinfo definition level would also disable evaluation tests which work fine, hence, only the specified gradient evaluation tests have been disabled within the test method.

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -16723,6 +16726,11 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
            dtypes=all_types_and(torch.float16, torch.bfloat16),
            supports_out=True,
            sample_inputs_func=sample_inputs_index_reduce),
+    *(OpInfo('index_reduce',
```
[the remark this answers is not in the document]
Comment (pearu) [reply torch/testing/_internal/common_methods_invocations.py:16729]: No reason other than the non-variant case represents the op with the reduction type `None` (which corresponds to `prod` as the default reduction type). I'll remove the non-variant case.

[the remark this answers is not in the document]
Comment (pearu) [reply torch/testing/_internal/common_methods_invocations.py:16729]: Fixed.

File: test/inductor/test_torchinductor_opinfo.py (hunk 1 above)
Comment (pearu) [reply test/inductor/test_torchinductor_opinfo.py]: Update: I have eliminated the if-block and moved the failing tests descriptions to index_reduce opinfo definitions.

Comment (pearu) [reply test/inductor/test_torchinductor_opinfo.py:526]: Update 2: Unfortunately, the if-block had to be restored. See the comment under the if-block.

File: test/test_proxy_tensor.py [hunk 3]
```diff
@@ -1868,6 +1868,12 @@ def f(t):
 
     # AssertionError: Tensor-likes are not close!
     skip('empty_strided', '', device_type='cpu'),
+
+    # RuntimeError: Cannot call numel() on tensor with symbolic sizes/strides
+    skip('index_reduce', 'prod'),
```
Comment (peterbell10) [inline test/test_proxy_tensor.py:1873]: These should be `xfail` rather than `skip`.

I guess these failures are expected if there's no meta impl, but do you know why there wasn't an xfail for the non-variant opinfo?

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -5232,6 +5232,8 @@ def make_idx(n, m):
     reduces = ('prod', 'mean', 'amin', 'amax')
 
     for shape, include_self, reduce in product(shapes, include_selfs, reduces):
+        if op_info.variant_test_name and op_info.variant_test_name != reduce:
```
Comment (peterbell10) [inline torch/testing/_internal/common_methods_invocations.py:5235]: NIT: You can remove this and just set `reduce = op_info.variant_test_name`

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -5255,9 +5257,10 @@ def make_idx(n, m):
         src = torch.tensor([[2, 0], [0, 0], [2, 3], [2, 2]], dtype=dtype, device=device, requires_grad=requires_grad)
         idx = torch.tensor([0, 1, 2, 0], dtype=torch.long, device=device)
 
-        yield SampleInput(input,
-                          args=(0, idx, src, 'prod'),
-                          kwargs={'include_self': True})
+        if not op_info.variant_test_name or op_info.variant_test_name == 'prod':
```
Comment (peterbell10) [inline torch/testing/_internal/common_methods_invocations.py:5260]:
```suggestion
        if reduce == "prod":
```

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -15907,6 +15910,8 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
                     skips=(
                         DecorateInfo(unittest.expectedFailure, 'TestNormalizeOperators', 'test_normalize_operator_exhaustive'),
                         DecorateInfo(unittest.expectedFailure, 'TestJit', 'test_variant_consistency_jit',),
+                        # NYI: The operator 'aten::rsub.Tensor' is not currently implemented for the MPS device
+                        DecorateInfo(unittest.expectedFailure, 'TestConsistency', 'test_output_grad_match'),
```
Comment (peterbell10) [inline torch/testing/_internal/common_methods_invocations.py:15914]: How are these MPS skips related to this PR?

File: test/inductor/test_torchinductor_opinfo.py (hunk 1 above)
Comment (peterbell10) [reply test/inductor/test_torchinductor_opinfo.py:526]: Okay, makes sense that we just weren't hitting the failure before. Can you move this special case into `inductor_override_kwargs` though with the setting `check_gradient=False` and add a TODO about the backward failing the test?

File: test/test_proxy_tensor.py (hunk 3 above)
Comment (pearu) [reply test/test_proxy_tensor.py:1873]: > These should be `xfail` rather than `skip`.

Fixed.

> I guess these failures are expected if there's no meta impl, but do you know why there wasn't an xfail for the non-variant opinfo?

yes:
(i) the non-variant opinfo emitted samples for all reduction types
(ii) `_test_make_fx_helper` produces only 5 samples when `out=True` which is the case for `test_make_fx_symbolic_exhaustive_out`.

The first 5 samples contain scalar inputs but the failure requires a non-scalar input. When enabling all other samples by disabling the if-block in https://github.com/pytorch/pytorch/blob/9dfeec9cdc246a6a003dff4b6ba0a5ceb60613f1/test/test_proxy_tensor.py#L2007, the failure can be reproduced on the main branch as well.

That said, the xfail instances is now moved to `out_symbolic_tensor_failures` list.

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (pearu) [reply torch/testing/_internal/common_methods_invocations.py:5235]: Nice catch!

File: torch/testing/_internal/common_methods_invocations.py (hunk 6 above)
Comment (pearu) [reply torch/testing/_internal/common_methods_invocations.py:15914]: These were related when trying to fix a suspicious code like
https://github.com/pytorch/pytorch/blob/9dfeec9cdc246a6a003dff4b6ba0a5ceb60613f1/test/test_mps.py#L195
(expected `key = op.name + '.' + op.variant_test_name if op.variant_test_name else op.name`) but that triggered too many failures in test_mps.py that fixes are OT to this PR. I can undo the MPS related skips but I believe these represent real issues, although, OT here.

File: test/inductor/test_torchinductor_opinfo.py (hunk 1 above)
Comment (pearu) [reply test/inductor/test_torchinductor_opinfo.py:526]: Done.

File: torch/testing/_internal/common_methods_invocations.py (hunk 5 above)
Comment (pearu) [reply torch/testing/_internal/common_methods_invocations.py:5260]: Fixed.

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (pearu) [reply torch/testing/_internal/common_methods_invocations.py:5235]: Done.

File: torch/testing/_internal/common_methods_invocations.py (hunk 6 above)
Comment (peterbell10) [reply torch/testing/_internal/common_methods_invocations.py]: I agree it's a weird way to format the key, but I don't think it's a problem and all the skips that you added are already listed in `test_mps.py` so I don't think we need to repeat them in the `OpInfo`.

Comment (pearu) [reply torch/testing/_internal/common_methods_invocations.py]: Not sure why the skips listed in test_mps.py were not effective at the time of testing... but I have now removed these from opinfos.

Comment (pearu) [reply torch/testing/_internal/common_methods_invocations.py]: After removing the skips, there are several MPS device related failures: https://hud.pytorch.org/pr/pytorch/pytorch/123375#23761295650

File: test/inductor/test_torchinductor_opinfo.py [hunk 7]
```diff
@@ -234,13 +239,15 @@ def format_op(op):
     "_upsample_bilinear2d_aa": {f16, f32, f64},
     "cholesky": {f32, f64},
     "multinomial": {f16, f32, f64},
-    "nn.functional.normalize": {f16},
     ("normal", "in_place"): {f16, f32, f64},
     ("normal", "number_mean"): {f16, f32, f64},
     "sparse.sampled_addmm": {f32, f64},
-    "to_sparse": {f16, f32, f64},
-    "torch.ops.aten._efficient_attention_forward": {f16, bf16, f32},
-    "torch.ops.aten._flash_attention_forward": {f16, bf16, f32},
```
Comment (peterbell10) [inline test/inductor/test_torchinductor_opinfo.py:243]: Looks like this is failing in CI

**Review by peterbell10 — APPROVED** [review]
